### PR TITLE
support sticky in IE

### PR DESCRIPTION
**Pull request by KeeyanGhoreshi**
This task is actually to fix the bug that the popup for the medications in the calendar don't show up sometimes.  The root cause is that IE doesn't support `position: sticky` and the html element that is used to anchor the popup can move off screen, so the popup does too.  To fix this issue I implemented a simple version of a sticky component for the medication items on the calendar.  It behaves almost identically to the css version of sticky.  It is a bit slower than the css version and has an issue where the table and the values will flicker out of view for a frame.  This can be a little jarring but is small enough that it can be ignored.  The table itself, or at least the header, already flickers when dragging through dates, but these changes for some reason cause it to affect the medication items as well.

## Submitter:

**Running the Application**

- [ ] Manually tested in Chrome (primary browser for testing)
- [ ] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [ ] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [ ] Branch implements task as expected
- [ ] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [ ] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [ ] This pull request describes why these changes were made
- [ ] Recognizes any potential shortcomings/bugs in the description 
- [ ] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [ ] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [ ] Document any new APIs/extension points
- [ ] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [ ] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [ ] Code is commented

**Tests**

- [ ] Existing tests passed
- [ ] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by Dtphelan1** [comment]
[DP1] - Got some linting errors that need resolving.

File: src/timeline/Timeline.scss [hunk 1]
```diff
@@ -112,7 +112,7 @@ $weekend: rgba(250, 246, 225, 0.5);
     .rct-item-content {
       position: sticky;
       position: -webkit-sticky;
-      left: 0px;
+
```
Comment (Dtphelan1) [inline src/timeline/Timeline.scss]:
```suggestion
```

File: src/timeline/TimelineEventsVisualizer.jsx [hunk 2]
```diff
@@ -20,13 +20,12 @@ class TimelineEventsVisualizer extends Visualizer {
         const { conditionSection, tdpSearchSuggestions, highlightedSearchSuggestion, isWide } = this.props;
         const items = this.createItems(conditionSection, tdpSearchSuggestions, highlightedSearchSuggestion);
         const groups = this.createGroupsForItems(this.getMaxGroup(items));
-
+        this.stickySupport = true;
```
Comment (Dtphelan1) [inline src/timeline/TimelineEventsVisualizer.jsx]: [DP1] Do we want to have this use some of the utilities defined in Slate for checking what Browser we're using (in `src/lib/slate/constants/environment.js`)? On the one hand that would allow us to actually use regular sticky on other browsers; on the other hand, I don't know we can reasonably expect `window.navigator` to be defined when we launch the app via Smart-on-FHIR. @rdingwell Do you  know anything about how/if the Smart-on-FHIR launcher would scramble any information stored in `window.navigator`?

Comment (KeeyanGhoreshi) [reply src/timeline/TimelineEventsVisualizer.jsx]: The sticky support is acquiesced from the CSS computed style, since unsupported "position" defaults to "static".  So this solution lets us use regular sticky on other browsers anyway.  Though if we can detect browsers that will probably be a more comprehensive, less hacky solution.

Comment (Dtphelan1) [reply src/timeline/TimelineEventsVisualizer.jsx]: Ehhh I think the way we do this now is fine. I retract this request.

File: src/timeline/TimelineEventsVisualizer.jsx [hunk 3]
```diff
@@ -274,6 +316,23 @@ class TimelineEventsVisualizer extends Visualizer {
         });
     }
 
+    updateMedicationItems() {
```
Comment (Dtphelan1) [inline src/timeline/TimelineEventsVisualizer.jsx]: [DP1] This should be an arrow function.

File: src/timeline/TimelineEventsVisualizer.jsx [hunk 4]
```diff
@@ -60,6 +59,27 @@ class TimelineEventsVisualizer extends Visualizer {
         };
     };
 
+    componentDidMount() {
+        window.addEventListener('load', this.handleLoad);
+    }
+
+    componentDidUpdate() {
+        if (!this.stickySupport) {
+            this.updateMedicationItems();
```
Comment (Dtphelan1) [inline src/timeline/TimelineEventsVisualizer.jsx]: [DP2] We should use lodash to throttle this somewhat. The performace is super choppy on my IE instance, and I anticipate it'll be marginally better with a throttled call here (by something like 300ms ish?)

File: src/timeline/TimelineEventsVisualizer.jsx (hunk 2 above)
Comment (Dtphelan1) [reply src/timeline/TimelineEventsVisualizer.jsx]: lol sorry this seecond comment was made before seeing your thoughtful response.  My bad!

Comment (noranda) [reply src/timeline/TimelineEventsVisualizer.jsx]: What about using a util? Something like:
```
// utils/supportsSticky
const supportsSticky = (() => {
  const node = document.createElement('div');
  node.style.position = 'sticky';
  return node.style.position === 'sticky';
})()

export default supportsSticky;```

File: src/timeline/TimelineEventsVisualizer.jsx (hunk 4 above)
Comment (KeeyanGhoreshi) [reply src/timeline/TimelineEventsVisualizer.jsx]: throttling it will make the text jump, it will only move once every 300ms.  Is that what we want?

**Comment by Dtphelan1** [comment]
I know it’s a not ideal... IMO, compared to the choppiness I think text-jumping is better than text-twitching. Open to opinions though, I’m not too set in my opinion. I’ll send a Gif tomorrow to show you what I’m seeing.   
  

  
  

  
  
>

**Comment by noranda** [comment]
[NB1] There is inconsistency between IE and Chrome with how the timeline now behaves. In Chrome, it displays a 1 year timeline showing the correct labels:
Screen Shot 2019-09-13 at 9 04 33 AM
In IE, on initial reload it displays a 1 year timeline with missing labels:
Screen Shot 2019-09-13 at 9 03 23 AM
In IE, after hovering over an element in the timeline will it then switch to a 3 year timeline showing the correct labels:
Screen Shot 2019-09-13 at 9 04 13 AM

This branch also needs to be rebased.

**Comment by KeeyanGhoreshi** [comment]
I can't seem to recreate the bug.  I'm using IE11 in a VM, I navigate to /pilot1 and wait for the page to load and then scroll down to the timeline and the labels are all where they should be on the 1 year display.  Hovering over the elements works fine as well.  Am I missing something?

**Comment by noranda** [comment]
@KeeyanGhoreshi I was using BrowserStack where I saw this. Perhaps someone on an actual Windows machine can test it?

**Comment by KeeyanGhoreshi** [comment]
The latest commit fixes quite a few bugs too numerous to name, but overall it should work in almost every normal working case.  Though it doesn't address Noranda's bug, since I still can't seem to reproduce it, everything worked fine on the demo machine in the lab, which is a windows machine.

**Comment by noranda** [comment]
@KeeyanGhoreshi Can you rebase this and fix the failed builds?

**Comment by noranda** [comment]
[NB2] The app breaks on the `/ccp` endpoint.

Screen Shot 2019-09-19 at 10 57 05 AM

**Comment by KeeyanGhoreshi** [comment]
Recent commit should deal with NB2

**Comment by noranda** [comment]
[NB1] and [NB2] are resolved, however, there is a failing test now and this needs to be rebased again.

**Comment by KeeyanGhoreshi** [comment]
the rebase should deal with the broken test

**Comment by noranda** [comment]
Looks good now!

**Comment by Dtphelan1** [comment]
The sticky behavior works as expected... however now that all the bugs have been ironed out of this new approach I'm not convinced that this is more performant than the original approach. I'd like to talk a little bit about this in the lab when you get in, but if it's not bringing a lot else to the table then maybe we want to consider.... dare I say.... reverting back to before our big brain play.

**Comment by Dtphelan1** [comment]
On second test, this looks wonderful. Great work Keeyan!

**Comment by Dtphelan1** [comment]
Once this is rebased again we'll merge.

**Comment by Dtphelan1** [comment]
This appears to work after the rebase, so I'm good to merge.
